### PR TITLE
feat(composer): support constraints

### LINKED ISSUE #6945: [Composer] Lock file maintenance bumps incompatible dependency
I have an issue with the weekly lock file maintenance for Composer. One of my dependency has `"ocramius/proxy-manager": "^2.8.0"` as a dependency. The weekly lockfile maintenance bumps the dependency from 2.8.0 to 2.9.0 in the lockfile.

```
             "name": "ocramius/proxy-manager",
-            "version": "2.8.0",
+            "version": "2.9.0",
             "source": {
                 "type": "git",
                 "url": "https://github.com/Ocramius/ProxyManager.git",
-                "reference": "ac1dd414fd114cfc0da9930e0ab46063c2f5e62a"
+                "reference": "9475248b6521f3029b973c4cd9aad31022bfd793"
             },
             "dist": {
                 "type": "zip",
-                "url": "https://api.github.com/repos/Ocramius/ProxyManager/zipball/ac1dd414fd114cfc0da9930e0ab46063c2f5e62a",
-                "reference": "ac1dd414fd114cfc0da9930e0ab46063c2f5e62a",
+                "url": "https://api.github.com/repos/Ocramius/ProxyManager/zipball/9475248b6521f3029b973c4cd9aad31022bfd793",
+                "reference": "9475248b6521f3029b973c4cd9aad31022bfd793",
                 "shasum": ""
             },
             "require": {
+                "composer-runtime-api": "^2.0.0",
                 "laminas/laminas-code": "^3.4.1",
-                "ocramius/package-versions": "^1.8.0",
                 "php": "~7.4.1",
                 "webimpress/safe-writer": "^2.0.1"
             },
```

Problem is, version 2.9.0 that was released doesn't work with Composer 1.0, but only with Composer 2.0 (that is still in alpha). As you can see, there is a new dependency to `"composer-runtime-api": "^2.0.0"` that is added. If I run a local `composer update`, the dependency stays on 2.8.0 (is even downgraded to 2.8.0 if I run it on the `renovate/lock-file-maintenance` branch).

Cf Ocramius/ProxyManager#621

### PR DESCRIPTION
**Pull request by viceice**
## Changes:
Extract composer version and use it as constraint if available.

## Context:

closes #6945

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added unit tests, or
- [x] Unit tests + ran on a real repository

### CODE REVIEW

**Comment by viceice** [comment]
tested here: https://github.com/viceice-tests/renovate-composer/pulls

File: lib/manager/composer/__snapshots__/artifacts.spec.ts.snap [hunk 1]
```diff
@@ -118,41 +118,7 @@ Array [
...
-        "NO_PROXY": "localhost",
-        "PATH": "/tmp/path",
-      },
-      "maxBuffer": 10485760,
-      "timeout": 900000,
-    },
-  },
-]
-`;
+exports[`.updateArtifacts() supports docker mode 1`] = `Array []`;
```
Comment (viceice) [inline lib/manager/composer/__snapshots__/artifacts.spec.ts.snap:121]: need to check this 😕

File: lib/manager/composer/__snapshots__/artifacts.spec.ts.snap [hunk 2]
```diff
@@ -73,7 +73,7 @@ Array [
 exports[`.updateArtifacts() returns null if unchanged 1`] = `
 Array [
   Object {
-    "cmd": "composer update --with-dependencies --ignore-platform-reqs --no-ansi --no-interaction --no-scripts --no-autoloader",
+    "cmd": "composer update --with-dependencies --ignore-platform-reqs --no-ansi --no-interaction",
```
Comment (rarkins) [inline lib/manager/composer/__snapshots__/artifacts.spec.ts.snap:76]: Is this related to the change?

**Comment by viceice** [comment]
ok, ready for another review. I can split this pr and extract some refactoring if required.

New pr's here: https://github.com/viceice-tests/renovate-composer/pulls

**Review by rarkins — APPROVED** [review]

File: lib/manager/composer/__snapshots__/artifacts.spec.ts.snap (hunk 2 above)
Comment (viceice) [reply lib/manager/composer/__snapshots__/artifacts.spec.ts.snap:76]: just to do some more code coverage. It's because i set the `trustLevel=high` on an existing test. 🙃

**Review by rarkins — APPROVED** [review]

**Comment by viceice** [comment]
Will update php docker image in near future, to support installing composer version on the fly, so we can contraint the php version too
